From bf79fe519d55a5fc6d31775164e0a4c5831a9020 Mon Sep 17 00:00:00 2001
From: Olivier Fourdan <ofourdan@redhat.com>
Date: Fri, 25 Oct 2024 14:04:17 +0200
Subject: [PATCH 3/5] compositor: Apply the window configuration

Use the "MetaWindow::configure" signal in mutter-16 to apply the GNOME
Kiosk window configuration.

(cherry picked from commit 30d0c4233f84234bd6d470fc42715593360e115a)
---
 compositor/kiosk-compositor.c | 47 +++++++++++++++++++++++++++++++++--
 1 file changed, 45 insertions(+), 2 deletions(-)

diff --git a/compositor/kiosk-compositor.c b/compositor/kiosk-compositor.c
index 37588da..650ac50 100644
--- a/compositor/kiosk-compositor.c
+++ b/compositor/kiosk-compositor.c
@@ -12,6 +12,7 @@
 #include <meta/keybindings.h>
 #include <meta/meta-context.h>
 #include <meta/util.h>
+#include <meta/meta-window-config.h>
 #include <meta/meta-window-group.h>
 
 #include <systemd/sd-daemon.h>
@@ -25,6 +26,7 @@
 #include "kiosk-window-tracker.h"
 #include "kiosk-shell-introspect-service.h"
 #include "kiosk-shell-screenshot-service.h"
+#include "kiosk-window-config.h"
 
 #include "org.gnome.DisplayManager.Manager.h"
 
@@ -48,6 +50,7 @@ struct _KioskCompositor
         KioskWindowTracker          *tracker;
         KioskShellIntrospectService *introspect_service;
         KioskShellScreenshotService *screenshot_service;
+        KioskWindowConfig           *kiosk_window_config;
 };
 
 enum
@@ -61,6 +64,8 @@ static guint signals[NUMBER_OF_SIGNALS] = { 0, };
 G_DEFINE_TYPE (KioskCompositor, kiosk_compositor, META_TYPE_PLUGIN)
 
 static void kiosk_compositor_dispose (GObject *object);
+static gboolean kiosk_compositor_wants_window_fullscreen (KioskCompositor *self,
+                                                          MetaWindow      *window);
 
 static void
 kiosk_compositor_dispose (GObject *object)
@@ -220,6 +225,39 @@ neuter_builtin_keybindings (KioskCompositor *self)
         }
 }
 
+static void
+kiosk_compositor_on_window_configure (MetaWindow       *window,
+                                      MetaWindowConfig *window_config,
+                                      gpointer          user_data)
+{
+        KioskCompositor *self = KIOSK_COMPOSITOR (user_data);
+        gboolean fullscreen;
+
+        if (!meta_window_config_get_is_initial (window_config)) {
+                g_debug ("KioskCompositor: Ignoring configure for window: %s",
+                         meta_window_get_description (window));
+                return;
+        }
+
+        g_debug ("KioskCompositor: configure window: %s", meta_window_get_description (window));
+
+        fullscreen = kiosk_compositor_wants_window_fullscreen (self, window);
+        meta_window_config_set_is_fullscreen (window_config, fullscreen);
+        kiosk_window_config_update_window (self->kiosk_window_config,
+                                           window,
+                                           window_config);
+}
+
+static void
+kiosk_compositor_on_window_created (MetaDisplay *display,
+                                    MetaWindow  *window,
+                                    gpointer     user_data)
+{
+        g_signal_connect (window, "configure",
+                          G_CALLBACK (kiosk_compositor_on_window_configure),
+                          user_data);
+}
+
 static void
 kiosk_compositor_start (MetaPlugin *plugin)
 {
@@ -250,6 +288,7 @@ kiosk_compositor_start (MetaPlugin *plugin)
         self->input_sources_manager = kiosk_input_sources_manager_new (self);
         self->app_system = kiosk_app_system_new (self);
         self->tracker = kiosk_window_tracker_new (self, self->app_system);
+        self->kiosk_window_config = kiosk_window_config_new ();
         self->introspect_service = kiosk_shell_introspect_service_new (self);
         kiosk_shell_introspect_service_start (self->introspect_service, &error);
         self->screenshot_service = kiosk_shell_screenshot_service_new (self);
@@ -265,6 +304,11 @@ kiosk_compositor_start (MetaPlugin *plugin)
                                                       self->cancellable,
                                                       KIOSK_OBJECT_CALLBACK (register_session),
                                                       NULL);
+
+        g_signal_connect_object (self->display, "window-created",
+                                 G_CALLBACK (kiosk_compositor_on_window_created),
+                                 self,
+                                 G_CONNECT_DEFAULT);
 }
 
 static void
@@ -372,9 +416,8 @@ kiosk_compositor_map (MetaPlugin      *plugin,
 
         window = meta_window_actor_get_meta_window (actor);
 
-        if (kiosk_compositor_wants_window_fullscreen (self, window)) {
+        if (meta_window_is_fullscreen (window)) {
                 g_debug ("KioskCompositor: Mapping window that does need to be fullscreened");
-                meta_window_make_fullscreen (window);
                 easing_duration = 3000;
         } else {
                 g_debug ("KioskCompositor: Mapping window that does not need to be fullscreened");
-- 
2.49.0

